File: app/src/main/kotlin/com/silent/vpn/policy/OtaGithubDiscovery.kt
package com.silent.vpn.policy

import com.google.gson.JsonParser

/**
 * OTA discovery that survives hive IP changes.
 * Landing already publishes [RELEASES_JSON_URL]; in-app check used to ask
 * `/api/updates/check` on the baked nip.io and died when that IP moved.
 */
object OtaGithubDiscovery {
    const val RELEASES_JSON_URL = "https://silentvpn3.github.io/releases.json"
    const val FETCH_TIMEOUT_MS = 8_000L

    fun landingKey(platform: String): String = when (platform.trim().lowercase()) {
        "android_tv", "tv" -> "android"
        "macos", "darwin" -> "mac"
        "windows", "win32" -> "pc"
        else -> platform.trim().lowercase()
    }

    sealed class Result {
        data class Available(
            val version: String,
            val filename: String,
            val size: Long,
            val downloadUrl: String,
        ) : Result()

        data class Current(val version: String) : Result()
        data object Unreadable : Result()
    }

    fun parse(json: String?, platform: String, currentVersion: String): Result {
        val root = runCatching {
            JsonParser.parseString(json ?: return Result.Unreadable).asJsonObject
        }.getOrNull() ?: return Result.Unreadable
        val key = landingKey(platform).ifBlank { return Result.Unreadable }
        val rowEl = root.get(key)
        if (rowEl == null || !rowEl.isJsonObject) return Result.Unreadable
        val row = rowEl.asJsonObject
        val latest = row.get("version")?.asString?.trim().orEmpty()
        val filename = row.get("filename")?.asString?.trim().orEmpty()
        val url = row.get("download_url")?.asString?.trim().orEmpty()
        if (latest.isEmpty() || filename.isEmpty() || !url.startsWith("http")) {
            return Result.Unreadable
        }
        if (!isNewer(latest, currentVersion)) {
            return Result.Current(latest)
        }
        return Result.Available(
            version = latest,
            filename = filename,
            size = row.get("size")?.asLong ?: 0L,
            downloadUrl = url,
        )
    }

    fun isNewer(latest: String, current: String): Boolean {
        val a = parseVersion(latest)
        val b = parseVersion(current)
        val n = maxOf(a.size, b.size)
        for (i in 0 until n) {
            val da = a.getOrElse(i) { 0 }
            val db = b.getOrElse(i) { 0 }
            if (da != db) return da > db
        }
        return false
    }

    private fun parseVersion(v: String): List<Int> =
        Regex("""\d+""").findAll(v).map { it.value.toInt() }.toList().ifEmpty { listOf(0) }
}

File: src/main/vpn/otaGithubDiscovery.js
'use strict'

const RELEASES_JSON_URL = 'https://silentvpn3.github.io/releases.json'
const FETCH_TIMEOUT_MS = 8_000

function parseVersion(v) {
  const parts = String(v || '').match(/\d+/g)
  return parts && parts.length ? parts.map((n) => parseInt(n, 10) || 0) : [0]
}

function isNewer(latest, current) {
  const a = parseVersion(latest)
  const b = parseVersion(current)
  const n = Math.max(a.length, b.length)
  for (let i = 0; i < n; i++) {
    const da = a[i] || 0
    const db = b[i] || 0
    if (da !== db) return da > db
  }
  return false
}

function landingKey(platform) {
  const p = String(platform || '').trim().toLowerCase()
  if (p === 'android_tv' || p === 'tv') return 'android'
  if (p === 'macos' || p === 'darwin') return 'mac'
  if (p === 'windows' || p === 'win32') return 'pc'
  return p
}

function parseGithubOta(raw, platform, currentVersion, arch) {
  let root
  try {
    root = typeof raw === 'string' ? JSON.parse(raw) : raw
  } catch {
    return { kind: 'unreadable' }
  }
  if (!root || typeof root !== 'object') return { kind: 'unreadable' }
  const key = landingKey(platform)
  let row = root[key]
  if (!row || typeof row !== 'object') return { kind: 'unreadable' }
  if (key === 'mac' && row.arches && typeof row.arches === 'object') {
    const archKey = arch === 'arm64' ? 'arm64' : 'x64'
    const slot = row.arches[archKey]
    if (!slot || typeof slot !== 'object' || !slot.version || !slot.download_url) {
      return { kind: 'unreadable' }
    }
    row = { ...row, ...slot }
  }
  const version = String(row.version || '').trim()
  const filename = String(row.filename || '').trim()
  const downloadUrl = String(row.download_url || '').trim()
  if (!version || !filename || !/^https?:\/\//i.test(downloadUrl)) {
    return { kind: 'unreadable' }
  }
  if (!isNewer(version, currentVersion)) {
    return { kind: 'current', version }
  }
  return {
    kind: 'available',
    version,
    filename,
    size: Number(row.size) || 0,
    download_url: downloadUrl,
    github_download_url: downloadUrl,
    available: true,
  }
}

/** GitHub first. Hive/tunnel only if there is no GitHub URL at all. */
function otaDownloadCandidates({
  githubUrl,
  vpnUp,
  tunnelOrigin = 'http://10.66.66.1:8000',
  hiveDownloadPath = '/api/updates/download/pc',
} = {}) {
  const out = []
  const seen = new Set()
  const add = (raw) => {
    const v = String(raw || '').trim()
    if (!v || seen.has(v)) return
    seen.add(v)
    out.push(v)
  }
  add(githubUrl)
  if (!out.length && vpnUp) {
    const path = String(hiveDownloadPath || '').trim() || '/api/updates/download/pc'
    const norm = path.startsWith('/') ? path : `/${path}`
    add(`${String(tunnelOrigin || '').replace(/\/$/, '')}${norm}`)
  }
  return out
}

module.exports = {
  RELEASES_JSON_URL,
  FETCH_TIMEOUT_MS,
  isNewer,
  landingKey,
  parseGithubOta,
  otaDownloadCandidates,
}
